### ISSUE TITLE
ALignment problem on back side

### ISSUE THREAD

**Issue by Terak8510** [pcb2gcode/pcb2gcode#656]
Git commit: latest-40-g5459c296
Boost: 107400
Gerbv: 2.7.0
Geos: Not installed

Using modified pcb2gcodeGUI with 'scribe-etch' method, scribe back side first, flip the board, scribe front side, drill, cut out.

config:
[pcb2gcodeGUI-current_scratch-etch.txt](https://github.com/pcb2gcode/pcb2gcode/files/10610083/pcb2gcodeGUI-current_scratch-etch.txt)

pictures front: 
![Screenshot_2023-02-04_18-25-23](https://user-images.githubusercontent.com/101630460/216798313-913591ab-80aa-4cf1-807e-21739ac97d4a.png)
![Screenshot_2023-02-04_18-35-23](https://user-images.githubusercontent.com/101630460/216798620-f7735972-792b-4336-a862-2de2173fb900.png)

pictures back:
![Screenshot_2023-02-04_18-29-29](https://user-images.githubusercontent.com/101630460/216798446-d0fe072a-0aa9-4213-aeb6-7bde9c2bd148.png)
![Screenshot_2023-02-04_18-34-21](https://user-images.githubusercontent.com/101630460/216798630-fc1d0d7d-d123-499a-a755-bed1b4aff7c4.png)

The is a little bit wander on the front side from the smaller drills flexing drilling a non piloted hole.
The back side is a bit of a dogs breakfast.

I havent upgraded git-src in a while, mabye thats it?

**Comment by Terak8510** [comment]
2 mods to the pcb2gcodeGUI source added i use:
![Screenshot_2023-02-04_18-58-15](https://user-images.githubusercontent.com/101630460/216799251-cf3f7d15-e486-421d-90e5-8d2fd9ea57d9.png)
![Screenshot_2023-02-04_18-58-53](https://user-images.githubusercontent.com/101630460/216799256-26e0ceb6-b684-4c1a-8135-5177a8529d84.png)

**Comment by eyal0** [comment]
When you flip it over, how do you make it line up?

**Comment by Terak8510** [comment]
At X=0 I have 2 index pins on oppisite ends of Y at edges of clad board, I tried  X and Y offsets but that does is offset ALL
 cnc jobs at once, front back drill cut etc.

**Comment by Terak8510** [comment]
might try the gcode on the 3d printer to see if it is consistant..

**Comment by Terak8510** [comment]
It could be candle v2.2, i have asked over there as well.
https://github.com/Schildkroet/Candle2/issues/41

**Comment by eyal0** [comment]
I don't use pcb2gcodeGUI, just pcb2gcode. Do you have test files that I can try?

**Comment by Terak8510** [comment]
k.
[the-zip-file.zip](https://github.com/pcb2gcode/pcb2gcode/files/10610291/the-zip-file.zip)
its in 7z format.

**Comment by eyal0** [comment]
I decompressed your files.  I also added a file called `millproject` to the directory and I put in the contents from [pcb2gcodeGUI-current_scratch-etch.txt](https://github.com/pcb2gcode/pcb2gcode/files/10610083/pcb2gcodeGUI-current_scratch-etch.txt) which you linked above.  Then I ran ~/pcb2gcode and it completed in 17 seconds on my computer.

I use `bCNC` to open the generated gcode and I got these two files from front and back:

![image](https://user-images.githubusercontent.com/109809/216803975-6d12e878-c00e-40a4-ae44-8421681507bd.png)

What you are describing should work.  Let me make sure that I understand:

First, you are placing your PCB on the CNC and you have two pins sticking out, one is at X=0,Y=<something> and the other is at X=0 and Y=<something else>.  And you have put on a copper clad that has two holes it in that match the two pins?  Also, when the CNC goes to X=0, it is exactly between these two pins?

If so, it should work.  I will show in bCNC:

![image](https://user-images.githubusercontent.com/109809/216804599-e494f82a-4f97-4a85-b220-097842e4834f.png)

I have drawn a measurement from the center to the center of the closest pad.  We see that, on the front, delta x is 2.19138mm and delta y is 2.16399mm.  And on the back, the measurements are the same except that delta x is negative.  This is because the flip is along axis 0.  (The measurements are not precisely the same because I'm limited by my ability to find the exact center of the pad with my mouse!)

So it seems like it should work.  Maybe your procedure is not what I described?  Can you describe in more details, perhaps with photos?

Also, I direct you to the wikipedia page on all of this:

https://github.com/pcb2gcode/pcb2gcode/wiki/Options:-Alignment-for-two-sided-PCB-and-offsets

**Comment by Terak8510** [comment]
sounds like you have possibly found the issue, the other piece of software is
Kicad, mabye the gerbers are not at the same position  with the exellon drl files?
here is a picture of my laseraxe 2417 cnc engraver with pins at X=0:
![20230204_234202](https://user-images.githubusercontent.com/101630460/216805304-5996df76-6fa8-4581-98b4-18cd17ab2d11.JPG)
left area of spoilboard is back, right area is front when i flip it over.

The board is purposely skewed to see if that mattered, it does not as long as the index pins are well fastened.

lower pin of copper board is at X=0, so is upper pin, just verified, powered down steppers and carved a straight line from upper index to lower index hole by manually turning Y knob and it drew a line right to the other index hole.
do you use kicad?

here is a screenshot of my plot and drill settings for that program:
![Screenshot_2023-02-04_22-46-27](https://user-images.githubusercontent.com/101630460/216805706-8230e093-7d93-4407-9c9d-195ecbfa40f2.png)
![Screenshot_2023-02-04_22-47-43](https://user-images.githubusercontent.com/101630460/216805711-cd49b7ef-46c5-49f2-b8c4-4a0dd25427b9.png)

**Comment by Terak8510** [comment]
currently googleing "kicad drill files not aligned with gerbers"

**Comment by Terak8510** [comment]
I turned up the stepper current on all 3 axis earlier today, it made no difference.
(thought it was skipping steps at high speed)
besides the inscribed pattern looks perfect.

**Comment by Terak8510** [comment]
I don't think it is a Kicad .grb/.drl issue, As the drill output is in perfect alignment
with the same 'side' of  the other gerber layer output.

it's just when i flip the board over.

**Comment by eyal0** [comment]
The drill file appears aligned to me.  In your images, looks like the back side is wrong in both x and y.  This seems strange.  Are you confident that your procedure for flipping over the board is maintaining the (0,0) correctly?  How could you test this...?

Well, you would jog your machine to x=0 and check that the pins that you are using for alignment when you flip are directly below the milling bit.  And it should be like this for both sides.

BTW, you are flipping only left to right, along the x axis, right?  You can't flip vertically, only horizontal.

I see the drilling coordinates in bCNC and they look perfect so I don't understand the issue.

**Comment by Terak8510** [comment]
Uh, I should have sent you the 'output' folder:
[cncfiles.zip](https://github.com/pcb2gcode/pcb2gcode/files/10610469/cncfiles.zip)

And currently running the .ngc files online through https://ncviewer.com/

**Comment by Terak8510** [comment]
meh, that online viewer only does 2 dimensional draws and cannot use multiple files.

I also noticed in Kicad, that the copper fill is larger than the outline.

**Comment by Terak8510** [comment]
I have found it,
The cnc engraver is not square on all 3 axis. (most likely)
I used a piece of photocard as a straight edge.

Unplugged and set X for middle -top of waste-board, 
Layed down a photocard, put an index pin there.
Hand turned Y from 0 to other limit, put an index pin there.

Drew a line from lower left edge of photocard on spoilboard.
Flipped over card on index holes and drew another line with pencil left to right.

I am measuring an error length of 3.21mm of the line ends of X.
Z axis is also not square as there is a little Y error of previous attempts.

This CNC has seen quite a few nasty bit crashes over the years.
closing.

**Comment by Terak8510** [comment]
I tightened every screw and checked for square on whole CNC frame.
Really not that far off, the photocard was probably not lined up right.
Even if the frame was out of alignment, Wouldn't the freshly drilled index holes 
take care of any alignment problems? 

Running one last scribe job to see if anything has changed.
![Screenshot_2023-02-05_21-28-06](https://user-images.githubusercontent.com/101630460/216891017-e9e2a6c0-155d-4903-9617-53cbfa3346a7.png)
At least now the holes are mostly inside the pads, the gantry still needs to be adjusted Y+/- ??? on one side.

**Comment by Terak8510** [comment]
Well i have realigned the CNC for squareness and resurfaced the spoilboard.
Still getting alignment errors even with big pads and holes in X direction.
![IMG_20230209_233832](https://user-images.githubusercontent.com/101630460/218058912-3d9ee4d7-dc3b-4be8-ac5e-d206184336cc.jpg)

Does not matter where i put the job on the spoilboard, the hole/pad is always
off to one side on X:
![Screenshot_2023-02-10_01-12-15](https://user-images.githubusercontent.com/101630460/218062916-27513467-9909-4d90-bd03-2b39bdc53fbd.png)

I can get around this by: 
drilling and milling front first,
flip board then run back job setting feed rate slow,
pause the job on first hole and turn X knob to compensate, then continue.
All the machine errors should have been eliminated when i resurfaced the spoilboard.

How difficult would it be to add --offset-side mill/drill option in the code?

**Comment by eyal0** [comment]
This must be a problem in your CNC, right?  I looked at the drilling and milling in bCNC and it's entirely accurate.  If you flip along the X axis, the holes should line up *precisely*.

Do you think that there remains an imprecision in your CNC or do you think that the error is in `pcb2gcode`.  If the latter, let's try to find definitive proof one way or the other.

**Comment by Terak8510** [comment]
Can confirm that it's NOT your code, I am stumped about why X alignment is off.
![Screenshot_2023-02-14_21-57-37](https://user-images.githubusercontent.com/101630460/218945468-a6244244-6514-4c85-99c1-12580ecfcd86.png)

Thanks for your patience.

**Comment by Terak8510** [comment]
This was solved by not setting --zero-start and enabling auxiliary axis in kicad.

**Comment by Terak8510** [comment]
Just for reference, I found the problem with why X/Y was not lining up when flipping the board.
There was a misalignment on the CNC, being X was not perpendicular with Y.
X was off by 0.8mm from one end to the other.
So when then board flipped, the error was now 1.6mm.
I loosened one side of X and used a straight edge between the rails of the X axis and moved the edge of the Y table right against the strait edge and re-tightened X rails.
Problem solved.
